streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java

```diff
@@ -523,24 +524,29 @@ public void shouldCountSessionWindows() throws Exception {
                 .windowedBy(SessionWindows.with(sessionGap))
                 .count()
                 .toStream()
-                .transform(() -> new Transformer<Windowed<String>, Long, KeyValue<Object, Object>>() {
-                        private ProcessorContext context;
-
-                        @Override
-                        public void init(final ProcessorContext context) {
-                            this.context = context;
-                        }
-
-                        @Override
-                        public KeyValue<Object, Object> transform(final Windowed<String> key, final Long value) {
-                            results.put(key, KeyValue.pair(value, context.timestamp()));
-                            latch.countDown();
-                            return null;
-                        }
-
-                        @Override
-                        public void close() {}
-                    });
+                .transform(new TransformerSupplier<Windowed<String>, Long, KeyValue<Object, Object>>() {
```

Comment:
Do we need to drop the SAM-converted function and go back to Supplier here, or is this just an artifact of testing? (If the latter, can you revert it?)

(also applies elsewhere)

Reply:
Unfortunately, the code does not compile with the SAM-converted function. I have not yet understood the reason. The compiler error is:
```
method transform in interface KStream<K,V> cannot be applied to given types;
                    .transform(() ->  new Transformer<Windowed<String>, Long, KeyValue<Object, Object>>() {
                    ^
  required: TransformerSupplier<? super Windowed<String>,? super Long,? extends KeyValue<? extends K1,? extends V1>>,String[]
  found: ()->new Tr[...]{ } }
  reason: cannot infer type-variable(s) K1,V1
    (argument mismatch; bad return type in lambda expression
      <anonymous Transformer<Windowed<String>,Long,KeyValue<Object,Object>>> cannot be converted to Transformer<Windowed<String>,Long,KeyValue<? extends K1,? extends V1>>)
  where K1,V1,K,V are type-variables:
    K1 extends Object declared in method <K1,V1>transform(TransformerSupplier<? super K,? super V,? extends KeyValue<? extends K1,? extends V1>>,String...)
    V1 extends Object declared in method <K1,V1>transform(TransformerSupplier<? super K,? super V,? extends KeyValue<? extends K1,? extends V1>>,String...)
    K extends Object declared in interface KStream
    V extends Object declared in interface KStream
```
When I use the old generics signature for `transform` it compiles. I will take a deeper look at it.

Reply:
Interesting.. is it because of the signature in `KStream#transform` for generics as `<? super K, ? super V, ..>` @vvcephei ?

Reply:
Specifically, the culprit is `? extends K1` and `? extends V1`. The type system won't go so far as to assume that if I return, say, Integer-valued results, then `V1` must be `Integer`. Because I didn't say I would _return_ a `V1`, just that whatever I return will _extend_ `V1`, and I never said what `V1` would actually be.

Unfortunately, it gets even less sensible from here... To solve that problem, it should be sufficient to fix V1 and K1, like this: `.<Integer, Integer>transform(...)`. But _that_ fails with a message that `Transformer<Windowed<String>, Long, KeyValue<Integer, Integer>>` can't be converted to `Transformer<Windowed<String>, Long, KeyValue<? extends Integer, ? extends Integer>>`.

I thought maybe it doesn't consider `Integer` to extend `Integer`, but `.<Number, Number>transform(...)` also doesn't work. I think what we're dealing with is just the insufficiency of Java's type system. I have heard several times that they deliberately stopped short of fully implementing the type system in exchange for performance, so there are some true type statements that nevertheless don't compile. Maybe this is one of them.

I think that we'd rather keep the ability to use lambda substitution for the supplier and give up the "maximum flexibility" on the return values for K1 and V1. Maybe you can just replace `? extends K1` with `K1` (and the same for `V1`) and re-attempt the experiments I went though before so we know what we're giving up?

Reply:
That sounds reasonable to me, especially given that the only caller of the constructor is always passing in a

```
TransformerSupplier<? super K, ? super V, KeyValue<K1, V1>> transformerSupplier
```

anyways.

Reply:
I did the experiments and the most flexible declaration that worked with the lambda is the following:
```
<K1, V1> KStream<K1, V1> transform(final TransformerSupplier<? super K, ? super V, ? extends KeyValue<K1, V1>> transformerSupplier,
                                   final String... stateStoreNames);
```
With this declaration, the `Transformer` may not return keys and values of subclasses of the key type and the value type of the result stream. Thus, the examples with `stream4` and `stream5` in @vvcephei 's experiments above do not compile with this declaration.

If nobody objects, I will proceed with the declaration above.

Removing only the `? extends` in front of `KeyValue` instead of removing `? extends` from `K1` and `V1` does also not compile when using lambdas. 

FYI, I created a minimal example of this rather strange compiler behavior and posted a question on stackoverflow [1].

[1] https://stackoverflow.com/questions/52445222/java-using-lambda-impedes-inference-of-type-variable

Reply:
That sounds good to me. Thanks for checking, @cadonna !

Reply:
Unfortunately, I had to use the following signatures in the end:

```
<K1, V1> KStream<K1, V1> transform(final TransformerSupplier<? super K, ? super V, KeyValue<K1, V1>> transformerSupplier,
                                   final String... stateStoreNames);

<K1, V1> KStream<K1, V1> flatTransform(final TransformerSupplier<? super K, ? super V, Iterable<KeyValue<K1, V1>>> transformerSupplier,
                                       final String... stateStoreNames);
```
The reason is that I got similar compile errors as before with the following signature for `flatTransform`:
```
<K1, V1> KStream<K1, V1> flatTransform(final TransformerSupplier<? super K, ? super V, ? extends Iterable<? extends KeyValue<K1, V1>>> transformerSupplier,
                                       final String... stateStoreNames);
```  
I tried a couple of alternatives, but the only one that allowed me to define a `TransformerSupplier` with an anonymous class and a lambda expression was the one without any wildcards in the generics for the output of the operator. To be consistent between `transform` and `flatTransform`, I also removed the wildcards in the generics for the output of `transform`. Now, `transform` has again the same signature as on the current trunk.

Lessons learned:
- Lambdas and anonymous classes have different representations in Java byte code. See [this answer](https://stackoverflow.com/a/52447112/6822244) to my stackoverflow question. I guess this is the reason for the differences in type variable inference.
- There would be ways to define a lambda for the more flexible signatures for `transform` and `flatTransform`, but they consists of casts and obscure generics declarations in the lambda (see [my stackoverflow question](https://stackoverflow.com/questions/52445222/using-lambda-impedes-inference-of-type-variable)). Additionally, it is a trial and error process to get the lambda right. Thus, something you do not want to expose to your user to.
- In Java 8, there exists [the `java.util.function.Supplier` functional interface](https://docs.oracle.com/javase/8/docs/api/java/util/function/Supplier.html) which is really similar to Kafka's `TransformerSupplier`. With the `java.util.function.Supplier` the more flexible signatures for `transform` and `flatTransform` seem to compile with lambdas (I have not thoroughly tested it, though). The drawback here is that defining the transformer supplier by means of an anonymous class does not seem to work anymore with the more flexible signature (also here, I have not invested too much time to try it out).

Reply:
Ah, well, I buy your reasoning, so I'm +1 for the interface you've settled on.
I doubt anyone really needs to subclass KeyValue anyhow.